ironfish/src/wallet/exporter/encoders/json.ts

```diff
@@ -1,38 +1,148 @@
-import { validateAccount } from '../../validator'
-import { AccountImport } from '../accountImport'
+import { generateKeyFromPrivateKey } from '@ironfish/rust-nodejs'
+import * as yup from 'yup'
+import { Assert } from '../../../assert'
+import { deserializeRpcAccountMultisigKeys } from '../../../rpc/routes/wallet/serializers'
+import { JSONUtils, YupUtils } from '../../../utils'
+import { ACCOUNT_SCHEMA_VERSION } from '../../account/account'
+import { AccountImport, validateAccountImport } from '../accountImport'
 import { AccountDecodingOptions, AccountEncoder, DecodeFailed } from '../encoder'
 
 export class JsonEncoder implements AccountEncoder {
   encode(value: AccountImport): string {
-    let createdAt = null
-    if (value.createdAt) {
-      createdAt = {
-        hash: value.createdAt.hash.toString('hex'),
-        sequence: value.createdAt.sequence,
-      }
-    }
-    return JSON.stringify({ ...value, createdAt })
+    const encoded = serializeAccountEncodedJSON(value)
+    return JSON.stringify(encoded)
   }
 
   decode(value: string, options?: AccountDecodingOptions): AccountImport {
-    let account: RpcAccountImport
-    try {
-      account = JSON.parse(value) as RpcAccountImport
-      const accountImport = deserializeRpcAccountImport(account)
+    const [raw, parseError] = JSONUtils.tryParse(value)
 
-      if (options?.name) {
-        accountImport.name = options.name
-      }
+    if (parseError) {
+      throw new DecodeFailed(`Invalid JSON: ${parseError.message}`, this.constructor.name)
+    }
+
+    const { result, error: schemaError } = YupUtils.tryValidateSync(
+      AccountEncodedJSONSchema,
+      raw,
+    )
+
+    if (schemaError) {
+      throw new DecodeFailed(`Invalid Schema: ${schemaError.message}`, this.constructor.name)
+    }
+
+    const account = deserializeAccountEncodedJSON(result)
 
-      validateAccount(accountImport)
-      return accountImport
-    } catch (e) {
-      throw new DecodeFailed(`Invalid JSON: ${(e as Error).message}`, this.constructor.name)
+    if (options?.name) {
+      account.name = options.name
     }
+
+    validateAccountImport(account)
+    return account
+  }
+}
+
+/**
+ * This is the type that represents all possible JSON encoded accounts
+ * from every supported possible backwards compatible account. DO NOT
+ * BREAK backwards compatability of this. If a field is optional, it
+ * means there exists an account in the wild that we exported that does
+ * not have that field.
+ */
+type AccountEncodedJSON = {
+  version?: number
+  name: string
+  viewKey?: string
+  incomingViewKey: string
+  outgoingViewKey: string
+  publicAddress: string
+  spendingKey: string | null
+  createdAt: { hash: string; sequence: number } | string | null
+  multisigKeys?: {
+    identity?: string
+    secret?: string
+    keyPackage?: string
+    publicKeyPackage: string
+  }
+  proofAuthorizingKey: string | null
+}
+
+const AccountEncodedJSONSchema: yup.ObjectSchema<AccountEncodedJSON> = yup
```

Comment:
This schema doesn't quite match the type AccountEncodedJSON, I'm guessing that's where some of the test failures are from

Reply:
Interestingly enough, I didn't actually change this. This is the old schema from RpcAccountImport. That means this is now catching failures that always existed now that we're actually checking the schema on import. That's a good thing. I'll investigate and make the schema match!